node/src/types/value_or_chunk.rs

```diff
@@ -0,0 +1,207 @@
+use std::{
+    fmt::{self, Display, Formatter},
+    marker::PhantomData,
+};
+
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use casper_hashing::{ChunkWithProof, Digest, MerkleConstructionError};
+use casper_types::bytesrepr::{self, Bytes, FromBytes, ToBytes, U8_SERIALIZED_LENGTH};
+use datasize::DataSize;
+
+/// Represents a value or a chunk of data with attached proof.
+/// Chunk with attached proof is used when the requested
+/// value is larger than [ChunkWithProof::CHUNK_SIZE_BYTES].
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub enum ValueOrChunk<V> {
+    /// Represents a value.
+    Value {
+        /// Serialized `V` type using `ToBytes` impl.
+        data: Bytes,
+        /// Phantom marker.
+        _phantom: PhantomData<V>,
+    },
+    /// Represents a chunk of data with attached proof.
+    ChunkWithProof(ChunkWithProof),
+}
+
+#[derive(Debug, PartialEq, Error, Eq)]
+/// Error returned when constructing an instance of `ValueOrChunk`.
+pub enum ChunkingError {
+    /// Merkle proof construction error
+    MerkleConstruction(MerkleConstructionError),
+    /// Serialization error
+    SerializationError(bytesrepr::Error),
+}
+
+impl Display for ChunkingError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            ChunkingError::MerkleConstruction(error) => f
+                .debug_tuple("MerkleConstructionError")
+                .field(&error)
+                .finish(),
+            ChunkingError::SerializationError(error) => {
+                f.debug_tuple("SerializationError").field(&error).finish()
+            }
+        }
+    }
+}
+
+impl<V> ValueOrChunk<V> {
+    /// Creates an instance of `ValueOrChunk::Value` if data is less than `max_chunk_size`
+    /// or a `ValueOrChunk::ChunkWithProof` if it is. In the latter case it will return only the
+    /// `chunk_index`-th chunk of the value's byte representation.
+    pub fn new(data: V, chunk_index: u64, max_chunk_size: usize) -> Result<Self, ChunkingError>
+    where
+        V: ToBytes,
+    {
+        let bytes = ToBytes::into_bytes(data).map_err(ChunkingError::SerializationError)?;
```

Comment:
The `AsRef<u8>` approach didn't pan out?

Reply:
Yes, for a couple of reasons: 
1. That would require having implemented `AsRef<[u8]>` for all types with which me might want to call this. We will be, most likely, calling `ValueOrChunk::new` with types that we already transfer over the wire - block effects, deploy (approvals), tries, etc. 
2. The choice of serialization (or turning `T` into bytes) has to be aligned with how it's deserialized in the fetcher since its hash might be used for id calculation/proof verification.

It seemed to me that using `ToBytes` will be the safest and most reasonable approach.

Reply:
Actually, now that I've learned that we use `bincode` to serialize data before sending, we should also use that same serialization mechanism for turning `T` into `Vec<u8>` - after all, we want to learn how big is the _wire representation_ of `T`.